### PR TITLE
TreeBehavior

### PR DESCRIPTION
**Pull request by skie**
Implemented left and right fields quoting in generated plain text queries.
Purpose of this pull request, make tree behavior working in case if field names used for left and right require field quoting, that's are necessery for some database drivers.

### CODE REVIEW

File: src/ORM/Behavior/TreeBehavior.php [hunk 1]
```diff
@@ -76,6 +77,23 @@ class TreeBehavior extends Behavior
     ];
 
     /**
+     * {@inheritDoc}
+     */
+    public function __construct(Table $table, array $config = [])
+    {
+        parent::__construct($table, $config);
+        $connection = $table->connection();
+        if (!is_null($connection)) {
```
Comment (dereuromark) [inline src/ORM/Behavior/TreeBehavior.php]: `!== null`

File: src/ORM/Behavior/TreeBehavior.php [hunk 2]
```diff
@@ -76,6 +77,23 @@ class TreeBehavior extends Behavior
...
+    {
+        parent::__construct($table, $config);
+        $connection = $table->connection();
+        if ($connection !== null) {
+            $driver = $connection->driver();
+            $this->config('leftField', $driver->quoteIdentifier($this->config('left')));
+            $this->config('rightField', $driver->quoteIdentifier($this->config('right')));
+        } else {
+            $this->config('leftField', $this->config('left'));
+            $this->config('rightField', $this->config('right'));
```
Comment (lorenzo) [inline src/ORM/Behavior/TreeBehavior.php]: This does not look right to me. Identifiers show not be quoted outside the query. I would rather see the same lines of the code that were already changed and replace accordingly with `new IdentifierExpression($leftField)`.

**Comment by lorenzo** [comment]
Aren't the fields quoted automatically anyway when activating th quote identifiers flag?

**Comment by skie** [comment]
expression written like this:

File: src/ORM/Behavior/TreeBehavior.php (hunk 2 above)
Comment (skie) [inline src/ORM/Behavior/TreeBehavior.php]: I agree with IdentifierExpression, but how would you use it in assignment expression? See my updateAll example in next comment.

**Comment by dereuromark** [comment]
@skie I reformatted your code, you forgot the code fence part (see edit).

**Comment by lorenzo** [comment]
@skie Then we can re-write those string into expression objects that can be quoted. Pre-quoting the strings is not the right way to go here

**Comment by skie** [comment]
@lorenzo We don't have way to build assignment expression with field both on left and right sides of the assignment.

File: src/ORM/Behavior/TreeBehavior.php (hunk 2 above)
Comment (markstory) [inline src/ORM/Behavior/TreeBehavior.php]: Won't configuration set at runtime run into the same quoting issue? Should we be quoting when expressions are built instead?

Comment (skie) [inline src/ORM/Behavior/TreeBehavior.php]: @markstory, So what do you propose. One possible supution would implement AssignmentExpression (class that will serve such sort of expressions). Possible alternative is have class that join mutiple expressions incuding plain strings into single string. In this case we will have two Identifier expressions with  '=' and '+1' strings.

Comment (skie) [inline src/ORM/Behavior/TreeBehavior.php]: @markstory if you asking why i put quoting into constructor then it is just becasue there a lot places where we really need quoting

**Comment by lorenzo** [comment]
closing in favor of https://github.com/cakephp/cakephp/pull/8027

**Comment by lorenzo** [comment]
@skie I created a new pull request with your commits and using `IdentifierExpression` as suggested

**Pull request closed without merging**